### DISCUSSION TITLE
[Feature Request] middleware for all API middleware routes (i.e `pages/api/_middleware.js`)

### DISCUSSION THREAD

**Discussion by garrettmac** [vercel/next.js#10325] [Ideas]
# Feature request
## problem statement

I’d like the ability to have certain functionality apply for all Api middleware routes. Similar to the concept of _document.js or _app.js.

I could use this for:
* logging all requests and responses in and out.
* Some transformation.
* Whitelisting or black listing requests.
* error tracking/handling

Really anything without having to import a function from a utility for each route.

## The solution you'd like

Something like a `pages/api/_middleware.js` that would default to current working directly and work up if I want to only opt-in or opt-out in some scenarios (so I could have `pages/api/foo/_middleware.js` only be middleware for `api/foo/*` endpoints if I wanted)

This would handle:
* Preflight request middleware support
* post response handling middleware
* error handling middleware

Maybe something like:

```jsx

import Cors from 'micro-cors'

// none of these ideally would allow sending response back before hitting a route. 
export async function preRequest(req, res) {
  const cors = Cors({
    allowMethods: ['GET', 'HEAD'],
  })
   return cors
// could wrap functions around each routes exported function, or do something with req or res like logging

}
// not sure if I like the naming
export async function postResponse(req, res) {
  // same as above
}

export async function onError(req, res) {
  // maybe not needed and could just capture in postResponse
}

```
Many other alternatives I’m sure but conceptually this feature would be nice.

Seen another issue where they just wanted in “onerror” I think this capture more use cases

**Comment by majelbstoat** [comment]
This would also be useful for supporting tracing. Would love to avoid having to add a custom server just to do logging/tracing etc. in a reasonable way.

**Comment by huang47** [comment]
This idea is awesome which also unlock the capabilities of permissioned routes which needs custom middlewares.

**Comment by ValentinH** [comment]
I really like the idea of a `_middleware.js` file and being able to place it at different levels. 

Another possible API I thought about:

```js
async function middleware(req, res, handler) {
  // here you can apply any logic you want and even return early (protected route?)
  console.log('before api handler')

  try {
    await handler(req, res)

    // after the handler, you can also do whatever you want (report metrics?)
    console.log('after handler')

  } catch (e) {
    // here you can implement a global error handler (error reporting)
    console.error({ msg: 'whatever' })
  }
}
export default middleware
```

**Comment by code-press** [comment]
Curious if something like this would work? https://github.com/oslabs-beta/connext-js
